mypy/constraints.py

```diff
@@ -290,12 +295,43 @@ def _infer_constraints(template: Type, actual: Type, direction: int) -> list[Con
             return handle_recursive_union(template, actual, direction)
         return []
 
+    if direction == SUBTYPE_OF and isinstance(actual, IntersectionType):
+        return any_constraints(
+            [
+                infer_constraints_if_possible(template, a_item, direction, preserve_typevars=True)
+                for a_item in actual.items
+            ],
+            eager=True,
+        )
+
+    if direction == SUPERTYPE_OF and isinstance(actual, IntersectionType):
+        items = simplify_away_incomplete_types(actual.items)
+        return any_constraints(
+            [infer_constraints(template, a_item, direction) for a_item in items], eager=True
+        )
+    if direction == SUBTYPE_OF and isinstance(template, IntersectionType):
+        return any_constraints(
+            [
+                infer_constraints_if_possible(t_item, actual, direction)
+                for t_item in template.items
+            ],
+            eager=False,
+        )
+    if direction == SUPERTYPE_OF and isinstance(template, IntersectionType):
+        return any_constraints(
+            [
+                infer_constraints_if_possible(t_item, actual, direction, preserve_typevars=True)
+                for t_item in template.items
+            ],
+            eager=False,
+        )
```

Comment:
```suggestion
    if isinstance(actual, IntersectionType):
        if direction == SUBTYPE_OF:
            return any_constraints(
                [
                    infer_constraints_if_possible(template, a_item, direction, preserve_typevars=True)
                    for a_item in actual.items
                ],
                eager=True,
            )
    
        if direction == SUPERTYPE_OF:
            items = simplify_away_incomplete_types(actual.items)
            return any_constraints(
                [infer_constraints(template, a_item, direction) for a_item in items], eager=True
            )

    if isinstance(template, IntersectionType):
        if direction == SUBTYPE_OF:
            return any_constraints(
                [
                    infer_constraints_if_possible(t_item, actual, direction)
                    for t_item in template.items
                ],
                eager=False,
            )
        if direction == SUPERTYPE_OF:
            return any_constraints(
                [
                    infer_constraints_if_possible(t_item, actual, direction, preserve_typevars=True)
                    for t_item in template.items
                ],
                eager=False,
            )
```

Reply:
I completely agree with this, but this bit of code is potentially refactorable, and extracting it like this will just make it harder. So I prefer to leave it as is